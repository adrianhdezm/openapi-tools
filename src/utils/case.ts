export function toPascalCase(str: string): string {
  return str
    .replace(/[^a-zA-Z0-9]+/g, ' ')
    .split(' ')
    .filter(Boolean)
    .map((p) => p.charAt(0).toUpperCase() + p.slice(1))
    .join('');
}

export function toCamelCase(str: string): string {
  const pascal = toPascalCase(str);
  return pascal.charAt(0).toLowerCase() + pascal.slice(1);
}
