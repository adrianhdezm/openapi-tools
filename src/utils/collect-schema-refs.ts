export function collectSchemaRefs(obj: unknown, refs: Set<string> = new Set()): Set<string> {
  if (!obj || typeof obj !== 'object') {
    return refs;
  }
  if ('$ref' in (obj as Record<string, unknown>) && typeof (obj as Record<string, unknown>)['$ref'] === 'string') {
    const ref = (obj as Record<string, unknown>)['$ref'] as string;
    const match = ref.match(/^#\/components\/schemas\/([^/]+)$/);
    if (match && match[1]) {
      refs.add(match[1]);
    }
  }
  for (const key in obj as Record<string, unknown>) {
    if (Object.prototype.hasOwnProperty.call(obj, key) && (obj as Record<string, unknown>)[key]) {
      collectSchemaRefs((obj as Record<string, unknown>)[key], refs);
    }
  }
  return refs;
}
