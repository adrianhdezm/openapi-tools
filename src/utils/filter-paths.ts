import { type OpenAPIV3 } from 'openapi-types';

export function filterOpenApiPaths(doc: OpenAPIV3.Document, pathNames: string[]): OpenAPIV3.Document {
  const filteredPaths: OpenAPIV3.PathsObject = {};
  for (const pathName of pathNames) {
    if (doc.paths && doc.paths[pathName]) {
      filteredPaths[pathName] = doc.paths[pathName];
    } else {
      console.warn(`\x1b[33m[openapi-tools] Warning:\x1b[0m Path "${pathName}" not found in the OpenAPI spec.`);
    }
  }

  // If no paths matched, return the original document
  const filtered: OpenAPIV3.Document = Object.keys(filteredPaths).length === 0 ? doc : { ...doc, paths: filteredPaths };
  return filtered;
}
