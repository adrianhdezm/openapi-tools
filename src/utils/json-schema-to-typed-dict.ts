import type { OpenAPIV3_1 as OpenAPI } from 'openapi-types';

export interface TypedDictResult {
  definition: string;
  typingImports: Set<string>;
}

export function convertToTypedDict(name: string, schema: OpenAPI.SchemaObject | OpenAPI.ReferenceObject): TypedDictResult {
  const typingImports = new Set<string>(['TypedDict']);
  const extraDefs: string[] = [];

  function toPascal(str: string): string {
    return str
      .replace(/[^a-zA-Z0-9]+/g, ' ')
      .split(' ')
      .filter(Boolean)
      .map((p) => p.charAt(0).toUpperCase() + p.slice(1))
      .join('');
  }

  function flatten(s: OpenAPI.SchemaObject | OpenAPI.ReferenceObject): { bases: string[]; schema: OpenAPI.SchemaObject | null } {
    if ('$ref' in s) {
      const match = s.$ref.match(/^#\/components\/schemas\/(.+)$/);
      return { bases: [match?.[1] ?? s.$ref], schema: null };
    }

    if ('allOf' in s && Array.isArray(s.allOf)) {
      const bases: string[] = [];
      const merged: OpenAPI.SchemaObject = { type: 'object', properties: {}, required: [] };
      for (const sub of s.allOf as Array<OpenAPI.SchemaObject | OpenAPI.ReferenceObject>) {
        const res = flatten(sub);
        bases.push(...res.bases);
        if (res.schema) {
          Object.assign(merged.properties!, res.schema.properties);
          merged.required = Array.from(new Set([...(merged.required ?? []), ...(res.schema.required ?? [])]));
        }
      }
      return { bases, schema: merged };
    }

    return { bases: [], schema: s };
  }

  function buildClass(className: string, s: OpenAPI.SchemaObject | OpenAPI.ReferenceObject): void {
    const { bases, schema: flat } = flatten(s);
    if (!flat || flat.type !== 'object') {
      return;
    }
    const props = flat.properties ?? {};
    const required = new Set(flat.required ?? []);
    const fields: string[] = [];
    const attrLines: string[] = [];
    for (const [key, value] of Object.entries(props)) {
      const typeStr = toType(value as any, `${className}${toPascal(key)}`);
      const desc = (value as any).description;
      if (required.has(key)) {
        typingImports.add('Required');
        fields.push(`    ${key}: Required[${typeStr}]`);
      } else {
        fields.push(`    ${key}: ${typeStr}`);
      }
      if (desc) {
        const descLines = String(desc).split('\n');
        attrLines.push(`${key}:`);
        for (const dl of descLines) {
          attrLines.push(`    ${dl}`);
        }
        attrLines.push('');
      }
    }
    if (fields.length === 0) {
      fields.push('    pass');
    }
    const baseList = bases.length > 0 ? `${bases.join(', ')}, ` : '';
    const header = [`class ${className}(${baseList}TypedDict, total=False):`];
    const docLines: string[] = [];
    if (flat.description) {
      docLines.push((flat.description as string).replace(/\n/g, ' '));
    }
    if (attrLines.length > 0) {
      if (attrLines[attrLines.length - 1] === '') attrLines.pop();
      if (flat.description) docLines.push('');
      docLines.push('Attributes:');
      for (const line of attrLines) {
        docLines.push(`    ${line}`);
      }
    }
    if (docLines.length > 0) {
      header.push(`    """\n    ${docLines.join('\n    ')}\n    """`);
    }
    extraDefs.push(`${header.join('\n')}\n${fields.join('\n')}`);
  }

  function toType(s: OpenAPI.SchemaObject | OpenAPI.ReferenceObject, className: string): string {
    if ('$ref' in s) {
      const match = s.$ref.match(/^#\/components\/schemas\/(.+)$/);
      const refName = match?.[1] ?? s.$ref;
      return `${refName}`;
    }

    if ('oneOf' in s && Array.isArray(s.oneOf)) {
      if (s.oneOf.length === 0) {
        typingImports.add('Any');
        return 'Any';
      }
      if (s.oneOf.length === 1) {
        return toType(s.oneOf[0] as any, className);
      }
      typingImports.add('Union');
      const parts = s.oneOf.map((sub, idx) => toType(sub as any, `${className}Option${idx}`)).join(', ');
      return `Union[${parts}]`;
    }

    if ('anyOf' in s && Array.isArray(s.anyOf)) {
      if (s.anyOf.length === 0) {
        typingImports.add('Any');
        return 'Any';
      }
      if (s.anyOf.length === 1) {
        return toType(s.anyOf[0] as any, className);
      }
      typingImports.add('Union');
      const parts = s.anyOf.map((sub, idx) => toType(sub as any, `${className}Option${idx}`)).join(', ');
      return `Union[${parts}]`;
    }

    if (s.enum) {
      typingImports.add('Literal');
      const values = s.enum.map((v) => JSON.stringify(v)).join(', ');
      return `Literal[${values}]`;
    }

    switch (s.type) {
      case 'string':
        return 'str';
      case 'number':
        return 'float';
      case 'integer':
        return 'int';
      case 'boolean':
        return 'bool';
      case 'array':
        if (!s.items) return 'List[Any]';
        typingImports.add('List');
        return `List[${toType(s.items, className)}]`;
      case 'object':
        if (s.properties && Object.keys(s.properties).length > 0) {
          buildClass(className, s);
          return className;
        }
        if (s.additionalProperties && typeof s.additionalProperties === 'object') {
          const valType = toType(s.additionalProperties as any, `${className}Additional`);
          return `dict[str, ${valType}]`;
        }
        typingImports.add('Any');
        return 'dict[str, Any]';
      default:
        typingImports.add('Any');
        return 'Any';
    }
  }

  const { bases, schema: flat } = flatten(schema);

  let definition = '';
  if (flat && flat.type === 'object') {
    const props = flat.properties ?? {};
    const required = new Set(flat.required ?? []);
    const fields: string[] = [];
    const attrLines: string[] = [];
    for (const [key, value] of Object.entries(props)) {
      const typeStr = toType(value as any, `${name}${toPascal(key)}`);
      const desc = (value as any).description;
      if (required.has(key)) {
        typingImports.add('Required');
        fields.push(`    ${key}: Required[${typeStr}]`);
      } else {
        fields.push(`    ${key}: ${typeStr}`);
      }
      if (desc) {
        const descLines = String(desc).split('\n');
        attrLines.push(`${key}:`);
        for (const dl of descLines) {
          attrLines.push(`    ${dl}`);
        }
        attrLines.push('');
      }
    }
    if (fields.length === 0) {
      fields.push('    pass');
    }
    const baseList = bases.length > 0 ? `${bases.join(', ')}, ` : '';
    const header = [`class ${name}(${baseList}TypedDict, total=False):`];
    const docLines: string[] = [];
    if (flat.description) {
      docLines.push((flat.description as string).replace(/\n/g, ' '));
    }
    if (attrLines.length > 0) {
      if (attrLines[attrLines.length - 1] === '') attrLines.pop();
      if (flat.description) docLines.push('');
      docLines.push('Attributes:');
      for (const line of attrLines) {
        docLines.push(`    ${line}`);
      }
    }
    if (docLines.length > 0) {
      header.push(`    """\n    ${docLines.join('\n    ')}\n    """`);
    }
    definition = `${header.join('\n')}\n${fields.join('\n')}`;
  } else {
    const typeStr = toType(schema, name);
    definition = `${name} = ${typeStr}`;
    if ((schema as any).description) {
      definition += `  # ${((schema as any).description as string).replace(/\n/g, ' ')}`;
    }
  }

  const combined = [...extraDefs, definition].filter(Boolean).join('\n\n');
  return { definition: combined, typingImports };
}
