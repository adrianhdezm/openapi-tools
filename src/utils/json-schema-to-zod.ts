import type { OpenAPIV3_1 as OpenAPI } from 'openapi-types';

export interface ZodResult {
  zodString: string;
  imports: Set<string>;
}

export function convertSchema(schema: OpenAPI.SchemaObject | OpenAPI.ReferenceObject): ZodResult {
  const imports = new Set<string>();

  function walk(s: OpenAPI.SchemaObject | OpenAPI.ReferenceObject): string {
    if ('$ref' in s) {
      const match = s.$ref.match(/^#\/components\/schemas\/(.+)$/);
      const name = match ? match[1] : s.$ref;
      imports.add(name);
      return `z.lazy(() => ${name})`;
    }

    if (s.enum) {
      const values = s.enum.map((v) => JSON.stringify(v)).join(', ');
      return `z.enum([${values}])`;
    }

    switch (s.type) {
      case 'string':
        return 'z.string()';
      case 'number':
      case 'integer':
        return 'z.number()';
      case 'boolean':
        return 'z.boolean()';
      case 'array':
        if (!s.items) return 'z.array(z.any())';
        return `z.array(${walk(s.items)})`;
      case 'object':
      default: {
        const props = s.properties ?? {};
        const required = new Set(s.required ?? []);
        const fields = Object.entries(props).map(([key, value]) => {
          let expr = walk(value as any);
          if (!required.has(key)) {
            expr += '.optional()';
          }
          return `${JSON.stringify(key)}: ${expr}`;
        });
        return `z.object({ ${fields.join(', ')} })`;
      }
    }
  }

  const zodString = walk(schema);
  imports.delete('');
  imports.delete((schema as any).title as string); // not needed
  return { zodString, imports };
}
