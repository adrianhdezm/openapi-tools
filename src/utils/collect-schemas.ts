import type { OpenAPIV3_1 as OpenAPI } from 'openapi-types';
import { collectSchemaRefs } from './collect-schema-refs.js';

export function collectSchemas(doc: OpenAPI.Document, refs: Set<string>): Record<string, OpenAPI.SchemaObject | OpenAPI.ReferenceObject> {
  const allSchemas = doc.components?.schemas ?? {};
  const collected: Record<string, OpenAPI.SchemaObject | OpenAPI.ReferenceObject> = {};
  const queue = Array.from(refs);
  const processed = new Set<string>();

  while (queue.length > 0) {
    const name = queue.pop()!;
    if (processed.has(name)) continue;
    const schema = allSchemas[name];
    if (schema) {
      collected[name] = schema;
      processed.add(name);
      collectSchemaRefs(schema, refs);
      for (const ref of refs) {
        if (!processed.has(ref) && !queue.includes(ref)) {
          queue.push(ref);
        }
      }
    }
  }

  return collected;
}
