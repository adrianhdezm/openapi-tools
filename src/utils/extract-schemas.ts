import type { OpenAPIV3_1 as OpenAPI } from 'openapi-types';
import { collectSchemaRefs } from './filter-paths.js';

export function extractSchemas(
  doc: OpenAPI.Document,
  prefixes: string[] | null
): Record<string, OpenAPI.SchemaObject | OpenAPI.ReferenceObject> {
  const allSchemas = doc.components?.schemas ?? {};
  if (!prefixes || prefixes.length === 0) {
    return allSchemas;
  }

  const refs = new Set<string>();
  for (const [pathName, pathItem] of Object.entries(doc.paths ?? {})) {
    if (prefixes.some((p) => pathName.startsWith(p))) {
      for (const method of Object.values(pathItem ?? {})) {
        if (typeof method === 'object' && method !== null) {
          collectSchemaRefs(method, refs);
        }
      }
    }
  }

  if (refs.size === 0) {
    return {};
  }

  const collected: Record<string, OpenAPI.SchemaObject | OpenAPI.ReferenceObject> = {};
  const queue = Array.from(refs);
  const processed = new Set<string>();

  while (queue.length) {
    const name = queue.pop()!;
    if (processed.has(name)) continue;
    const schema = allSchemas[name];
    if (schema) {
      collected[name] = schema;
      processed.add(name);
      collectSchemaRefs(schema, refs);
      for (const ref of refs) {
        if (!processed.has(ref) && !queue.includes(ref)) {
          queue.push(ref);
        }
      }
    }
  }

  return collected;
}
