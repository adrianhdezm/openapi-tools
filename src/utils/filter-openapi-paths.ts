import type { OpenAPIV3_1 as OpenAPI } from 'openapi-types';
import { collectSchemaRefs } from './collect-schema-refs.js';
import { collectSchemas } from './collect-schemas.js';

export type JsonValue = string | number | boolean | null | JsonObject | JsonValue[];
export type JsonObject = { [key: string]: JsonValue };

export function filterOpenApiPaths(doc: OpenAPI.Document, pathNames: string[]): OpenAPI.Document {
  const filteredPaths: OpenAPI.PathsObject = {};
  for (const pathName of pathNames) {
    if (doc.paths && doc.paths[pathName]) {
      filteredPaths[pathName] = doc.paths[pathName];
    } else {
      console.warn(`\x1b[33m[openapi-tools] Warning:\x1b[0m Path "${pathName}" not found in the OpenAPI spec.`);
    }
  }

  // If no paths matched, return the original document
  if (Object.keys(filteredPaths).length === 0) {
    console.warn('\x1b[33m[openapi-tools] Warning:\x1b[0m No paths matched the filter criteria. Returning the original OpenAPI document.');
    return doc;
  }

  // Collect all schema refs from the filtered paths
  const schemaRefs = new Set<string>();
  for (const path of Object.values(filteredPaths)) {
    for (const method of Object.values(path ?? {})) {
      if (typeof method === 'object' && method !== null) {
        collectSchemaRefs(method, schemaRefs);
      }
    }
  }

  const collectedSchemas = collectSchemas(doc, schemaRefs);

  const filtered: OpenAPI.Document = {
    ...doc,
    paths: filteredPaths,
    components: {
      ...(doc.components || {}),
      schemas: collectedSchemas
    }
  };

  return filtered;
}
