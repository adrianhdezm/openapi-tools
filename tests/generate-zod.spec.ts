import { describe, it, expect } from 'vitest';
import fs from 'fs';
import path from 'path';
import Handlebars from 'handlebars';
import ts from 'typescript';
import type { OpenAPIV3_1 as OpenAPI } from 'openapi-types';
import { convertSchema } from '../src/utils/json-schema-to-zod';
import { extractSchemas } from '../src/utils/extract-schemas';

const templateDir = path.resolve('templates/ts');
const schemaTemplate = Handlebars.compile(fs.readFileSync(path.join(templateDir, 'zod-schema.hbs'), 'utf8'));

describe('generate-zod', () => {
  it('generates a simple object schema', () => {
    const doc: OpenAPI.Document = {
      openapi: '3.1.0',
      info: { title: 't', version: '1' },
      paths: {},
      components: { schemas: { User: { type: 'object', properties: { id: { type: 'string' } }, required: ['id'] } } }
    };
    const schemas = extractSchemas(doc, null);
    const { zodString } = convertSchema(schemas.User as OpenAPI.SchemaObject);
    const content = schemaTemplate({
      schemas: [{ schemaName: 'userSchema', zodString }]
    });
    const result = ts.transpileModule(content, { compilerOptions: { module: ts.ModuleKind.ESNext } });
    expect(result.diagnostics?.length).toBe(0);
    expect(content).toMatchSnapshot();
  });

  it('generates enums and nested arrays', () => {
    const doc: OpenAPI.Document = {
      openapi: '3.1.0',
      info: { title: 't', version: '1' },
      paths: {},
      components: {
        schemas: {
          Status: { type: 'string', enum: ['on', 'off'] },
          Wrapper: { type: 'array', items: { $ref: '#/components/schemas/Status' } }
        }
      }
    };
    const schemas = extractSchemas(doc, null);
    const { zodString } = convertSchema(schemas.Wrapper as OpenAPI.SchemaObject);
    const content = schemaTemplate({
      schemas: [{ schemaName: 'wrapperSchema', zodString }]
    });
    const result = ts.transpileModule(content, { compilerOptions: { module: ts.ModuleKind.ESNext } });
    expect(result.diagnostics?.length).toBe(0);
    expect(content).toMatchSnapshot();
  });

  it('filters schemas by path prefixes', () => {
    const doc: OpenAPI.Document = {
      openapi: '3.1.0',
      info: { title: 't', version: '1' },
      paths: {
        '/users/list': {
          get: {
            responses: { '200': { description: 'ok', content: { 'application/json': { schema: { $ref: '#/components/schemas/User' } } } } }
          }
        },
        '/admin/data': {
          get: {
            responses: { '200': { description: 'ok', content: { 'application/json': { schema: { $ref: '#/components/schemas/Admin' } } } } }
          }
        }
      },
      components: {
        schemas: {
          User: { type: 'object', properties: { id: { type: 'string' } } },
          Admin: { type: 'object', properties: { secret: { type: 'string' } } }
        }
      }
    };
    const schemas = extractSchemas(doc, ['/users']);
    expect(Object.keys(schemas)).toEqual(['User']);
  });
});
