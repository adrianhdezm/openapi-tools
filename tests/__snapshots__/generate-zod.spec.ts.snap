// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate-zod > generates a simple object schema 1`] = `
"
export const User = z.object({
  id: z.string()
});
"
`;

exports[`generate-zod > generates enums and nested arrays 1`] = `
"import { Status } from './Status';\\n
export const Wrapper = z.array(z.lazy(() => Status));
"
`;
